File: CHUOND_4/4.19_FrameLayout_Java/app/src/main/java/com/example/myapplication/ui/MainActivity.java
package com.example.myapplication.ui;

import android.os.Bundle;
import android.widget.ListView;
import android.widget.Toast;

import androidx.activity.EdgeToEdge;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.content.res.AppCompatResources;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;
import androidx.lifecycle.ViewModelProvider;

import com.example.myapplication.R;
import com.example.myapplication.data.Country;
import com.example.myapplication.data.CountryRepository;
import com.example.myapplication.data.LocalDataSource;

public class MainActivity extends AppCompatActivity {
    private CountryAdapter mCountryAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        EdgeToEdge.enable(this);
        setContentView(R.layout.activity_main);
        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main), (v, insets) -> {
            Insets systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars());
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom);
            return insets;
        });
        setupViews();
        observeData();
    }

    private void observeData(){
        LocalDataSource localDataSource = new LocalDataSource(getResources());
        CountryRepository repository = new CountryRepository(localDataSource);
        CountryViewModel.Factory factory = new CountryViewModel.Factory(repository);
        CountryViewModel mViewModel =
                new ViewModelProvider(this, factory).get(CountryViewModel.class);
        mViewModel.loadCountries();
        mViewModel.getCountries().observe(this, mCountryAdapter::updateCountries);
    }

    private void setupViews(){
        ListView listCountry = findViewById(R.id.list_country);
        mCountryAdapter = new CountryAdapter();
        listCountry.setAdapter(mCountryAdapter);
        listCountry.setDivider(AppCompatResources.getDrawable(this, R.drawable.divider));
        listCountry.setOnItemClickListener((parent, view, position, id) -> {
            Country country = (Country) parent.getItemAtPosition(position);
            Toast.makeText(MainActivity.this, "Choose country: " + country.getName(), Toast.LENGTH_SHORT).show();
        });
    }
}

File: CHUOND_4/4.13_LiveData_Kotlin/app/src/main/java/com/example/myapplication/MainActivity.kt
package com.example.myapplication

import android.os.Bundle
import android.widget.Button
import android.widget.TextView
import androidx.activity.enableEdgeToEdge
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.lifecycle.ViewModelProvider
import com.google.android.material.textfield.TextInputEditText

class MainActivity : AppCompatActivity() {
    private lateinit var txtNewName: TextView
    private lateinit var nameViewModel: NameViewModel
    private lateinit var edtFullName: TextInputEditText
    private lateinit var btnChangeName: Button
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enableEdgeToEdge()
        setContentView(R.layout.activity_main)
        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main)) { v, insets ->
            val systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars())
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom)
            insets
        }
        setupViews()
        setupAction()
        setupViewModel()
    }

    private fun setupAction(){
        btnChangeName.setOnClickListener {
            val fullName = edtFullName.text.toString().trim()
            nameViewModel.updateName(fullName)
        }
    }

    private fun setupViews(){
        edtFullName = findViewById(R.id.edt_full_name)
        txtNewName = findViewById(R.id.txt_new_name)
        btnChangeName = findViewById(R.id.btn_change)
    }

    private fun setupViewModel(){
        nameViewModel = ViewModelProvider(this)[NameViewModel::class.java]
        nameViewModel.currentName.observe(this){
            fullName -> txtNewName.text = getString(R.string.text_result, fullName)
        }
    }
}